### ISSUE TITLE
Master 128 model seems to run a bit fast

### ISSUE THREAD

**Issue by BigEd** [mattgodbolt/jsbeeb#115]
Running from the pi.ssd pi calculation program, version 12, at
http://litwr2.atspace.eu/pi/pi-spigot-benchmark.html
I'm getting timings from JSBeeb which are somewhat faster than a real Master.
On my Master, for 1000 digits it reports 176.23
On JSBeeb it reports 158.74 or 158.75

More conveniently, computing 80 digits takes 1.18 on my Master, 1.07 on JSBeeb

I've attached v12 of the program - binary only - see the website for sources.
[pi.ssd.gz](https://github.com/mattgodbolt/jsbeeb/files/362321/pi.ssd.gz)

See also the discussion at http://stardot.org.uk/forums/viewtopic.php?f=2&t=11458

**Comment by BigEd** [comment]
A possible way to help debug which instruction timing might be wrong: kill the system timer, set up a user timer, capture the PC value when the timer fires. Binary search the timer value to find the opcode which is mistimed...  of course there might be a simpler way!

**Comment by richtw1** [comment]
Hey Ed,

The first thing I would try is to disable all IRQs except the 100Hz timer one, to see if that brings the timing closer to reality.  I suspect this isn't an instruction timing thing, but an unimplemented hardware thing which is "earlying out" under emulation, but requiring some extra IRQ handling time on real hardware.  Maybe the ADC? (which moved address on the Master).

Try *FX16,0 to get rid of ADC conversions, and something like

```
LDA #&7F:STA &FE4E:STA &FE6E
LDA #&C0:STA &FE4E   ; just Sys VIA T1
```

to get rid of all the VIA interrupts.

Also, do you have Econet or anything like that Ed, which might be generating extra NMIs?

**Comment by richtw1** [comment]
Just to get it down here, the only opcodes I can think of whose timing differs on a Master are

* JMP (ind)  - 1 cycle more
* ADC/SBC in decimal mode - 1 cycle more

I'm fairly sure the first case is already handled, don't know about decimal mode on the CMOS 6502, would have to look.  Doesn't look as if the test code uses it anyway.

**Comment by BigEd** [comment]
Hi Rich
  with those interventions, the timing on my Master comes to 1.16

  are you sure about all page-crossing timing adjustments?

**Comment by richtw1** [comment]
Ah, I've just remembered another one: there's a page-not-crossed optimization in ASL/ROL/LSR/ROR in CMOS so it takes a cycle less - I don't think that's supported in jsbeeb - but that would imply a slower than normal result under emulation.

Otherwise page-crossing should be correct - it's the same in both CPUs and we went to great lengths to make sure it was correctly implemented, so that the difficult VIA timer protection systems worked correctly!

I'm out of ideas then.

**Comment by BigEd** [comment]
Hmm. Warrants investigation?
BTW with your suggested interventions, JSBeeb completes 80 digits in 1.06

**Comment by richtw1** [comment]
This is the very definition of clutching at straws: do you have any sideways ROMs installed which might be hooking into the 100Hz poll service call?  That's a Master-only extension which would obviously be using extra CPU if it were doing anything.

**Comment by richtw1** [comment]
Also it'd be handy to know if there's any discrepancy when running on / emulating a BBC B.

**Comment by BigEd** [comment]
My B was poorly last time I checked but I can revisit it. (Tomorrow.) From published figures on that website I think jsbeeb was pretty close running a B - closer than this 10% discrepancy.

I don't have any special ROMs... but I should probably check. I think it would be cleanest to disable the system timer and use a user timer instead, then we know nothing rogue is running.

Is it possible my OS is different enough to make a difference? All it's doing is the VDU driver writing to screen. And checking for Escape. I'd be very surprised if it makes a substantial difference but it would make it a little more difficult to compare progress by using a timer. Timing the first VDU output might be the place to start, as I think we can intercept the VDU vector.

**Comment by BigEd** [comment]
Good news - my Model B is working at present. Also good news, it's cycle-exact against JSBeeb.

But I can confirm the Master is adrift. What I did was use some code at &100 to disable interrupts, save the system timer values, run the pi code, then save the new system timer values and re-enable interrupts. We don't get the elapsed time because the counters overflow, but a mismatch indicates an inexactitude in the emulation.

Changes to PI2100:
Make some space and add my calls:
15 F%=7
Line 80 change
CALLM%
to
CALL&104:CALLM%:CALL&112

Results on subtracting the two 16-bit values from
P.~!&100
after the program runs:

On a Model B in JSBeeb:
4 digits gives us 64369
8 digits gives us 65145
80 digits gives us 62784
which exactly matches a real B

On a Master in JSBeeb:
4 digits returns FD91AE5 which is 62708
8 digits returns A9615A2 which is 63204
80 digits returns 1A6519DC which is 137

Master:
4 digits gives 62697 
8 digits gives 63177
80 digits gives 64605

**Comment by BigEd** [comment]
(BTW my attempts there to subtract two values might be wrong because the timer reloads 9998, not 65535. But I think we still expect a cycle-accurate simulation to come up with the same values.)

I have a new plan: intercept the IRQ handler at IRQ1V and capture the first PC value off the stack. Also, of course, initialise the timer to a known value before starting. We can easily probe 65ms into the running of the program - if that's not far enough, we have to count invocations or save multiple PC values.

**Comment by BigEd** [comment]
I've binary-searched down to a single timer value which differs in the PC which it captures when it runs the spigot test. But the instructions in question when tested in isolation do not differ - so I'm quite confused. Here's the area of interest:
```
3008 E578    x  sbc 78
300A 267D   &}  rol 7D
300C 2A     *   rola 
300D C578    x  cmp 78
300F 9002       bcc 3013
3011 E578    x  sbc 78
3013 267D   &}  rol 7D
```

What I found is the real Master saving a PC value of 300C, which means the ROL 7D was interrupted, whereas the JSBeeb Master saves a PC value of 300D, meaning the ROL A was interrupted.

Any idea how this could happen if the instructions in isolation take the right time?

**Comment by richtw-epic** [comment]
I've just seen, we don't yet perform the extra cycle for JMP (ind).

https://github.com/mattgodbolt/jsbeeb/blob/master/6502.opcodes.js#L1055

It may just be as simple a fix as doing the TODO, e.g. changing line 1052 of 6502.opcodes.js to:

`ig.tick(is65C12 ? 4 : 3);`

Don't have time to try it now, but feel free to give it a try Ed!

**Comment by mattgodbolt** [comment]
Am at work (first day back after holiday) and trying best to ignore these amazing emails...sounds like you're on the right track though! If @BigEd you don't have a chance to try later I'll give this a go myself tonight!

**Comment by BigEd** [comment]
I'll be off this adventure for at least 14 hours now, so it would be great if either of you could get there before I do. It's a bit mystifying.

**Comment by mattgodbolt** [comment]
Just made the fix to the (ind)...no difference in timings sadly: I'm still seeing 1.07 for 80 digits. I haven't applied the other changes @BigEd has made (yet) though, but I can't see how those will help in this case.

I wonder if anything in *CONFIGURE-land could cause the discrepancies too (though once the IRQ/etc patches are applied only NMIs can get in the way...)

**Comment by BigEd** [comment]
I don't think there can be NMIs ... but I suppose I could put a BRK at &0D00 or a branch-to-self. Perhaps we should find another Master to check that my machine is acting as it should.

It would be good, if not too hard, to fix the other bug with the RMW,X addressing modes, because although it goes in the opposite direction it's possible it could be masking something.

**Comment by mattgodbolt** [comment]
I also experimented with making both () and (abs) take hundreds of cycles:
it slows down a little but nowhere enough to explain the discrepancy. So I
don't think either address mode is to do with the difference. Will keep
digging...

**Comment by richtw1** [comment]
Try at least *TAPE before running it, to hopefully clear out NMIs.  I'm really puzzled by this one, as the instruction timings are largely shared between the two CPU emulations.

**Comment by richtw1** [comment]
Also it'd be good to know if the timing is any better if the calls to JSR OSWRCH are removed.

**Comment by BigEd** [comment]
I'm very confused now... it looked for an instant like *TAPE fixed it - I'm now getting the right timings from my Master. But it's worse than that - I'm getting the right timings even if I don't use *TAPE. So, more investigation needed on my side, but it's now clear that a real Master can match JSBeeb, even if it might find itself in a mode where it doesn't.
Probably I'm too tired now to investigate further, even if I had an idea!
Sorry about this, but thanks for the TAPE suggestion.

**Comment by mattgodbolt** [comment]
It's all good stuff! Thanks so much for your diligent efforts to get this
stuff right!!

**Comment by BigEd** [comment]
(BTW, when I say Master matching the Beeb, I mean in the gross centisecond timing of the spigot program. There's still the one-cycle interrupt timing discrepancy.)

**Comment by mattgodbolt** [comment]
Hey Ed, can you elaborate (and/or post your SSD) of the one-cycle IRQ
timing discrepancy? I think I'm missing a step somewhere in reproducing it
(or I've not had enough coffee! :))

Thanks mate, Matt

**Comment by BigEd** [comment]
OK, see attached. It's probably worth revisiting this idea: initialise the timer to get into a known state, set up an ISR to catch the PC next time the interrupt fires, then run some code. Compare JSBeeb and a physical machine, see if the PC ever differs between the two. If it does, change the initial timer value a binary search to narrow down the first mis-timed interrupt. Then scratch your head.

In the ssd, you'll find some EXECable sources:
PIT - a version of the BASIC wrapper which has such a gadget. It's initialising to &0800, but the binary search converged on the first fail at &07C1
test2 - simple single-instruction timer
TIM1 - timing a composite of the RMW instructions
[pi-mod.ssd.gz](https://github.com/mattgodbolt/jsbeeb/files/374032/pi-mod.ssd.gz)

**Comment by mattgodbolt** [comment]
Thanks so much! Will definitely take a look on the train ride home tonight!

**Comment by BigEd** [comment]
I should add - the testcase for PIT is to ask for 4 digits, in mode 7.

**Comment by BigEd** [comment]
A datapoint on the centisecond timing: hoglet reported his timings in the stardot thread for 1000 digits as
Master 128 - 160.51s
Master 128 with *FX 16,0 - 158.88s
In JSBeeb I see 158.77 normally and 157.97 with *FX 16,0

**Comment by mattgodbolt** [comment]
I confirm that with the PIT test and 7C1 the interrupt happens at 300C on jsbeeb; but without a master to cross check I'm a bit stuck!

Kieran (I think you're watching this...); are you able to follow along too and maybe repro on a master ?

**Comment by BigEd** [comment]
I think I've found a problem with my Master: it has some kind of sticky key problem, possibly the caps lock or shift lock. I'd thought that was irrelevant except for being inconvenient, but I see that timing a simple Basic loop in JSBeeb, holding down caps lock does slow the machine down by 8% or 9%. So, a sticky key can cause a slow Master, which throws out my centisecond timings.
But hoglet's less variant timings are probably still a valid measurement, and the interrupt timing discrepancy is probably still real.

**Comment by mattgodbolt** [comment]
Fantastic find!!! Weird but sort of makes sense!

On the IRQ timing thing, can you confirm that I'm doing it right now: I see
an interrupt on 300C now when I put the counter to 7C1; is that now the
same as your Master?

**Comment by BigEd** [comment]
Yes, 300C is the result on the master. JSBeeb was previously reporting 300D - are you saying you've fixed it already?
This was my binary search, reporting the stack and PC value from the peek:
0800, master says F9 305D, jsBeeb says 305F
0400, master says F9 2182. jsbeeb B says same.
0600, master says F9 21C7. jsbeeb B says same.
​0700, master says F9 2401, jsbeeb B says same. As does jsMaster.
0780, master says F9 2FB9, jsbeeb B says same.
07C0, master says F9 300C, jsbeeb B says same. As does jsMaster.
07E0, 3032, 3033 - so there's a discrepancy at last!
07D0 301F, 3021 - a bigger discrepancy...
07C8 3015, 3016 - still discrepant
07C4 300F, 3013 - still wrong
07C2 300D, 300F - still wrong
07C1 300C, 300D - still wrong, and that's the first fail. jsMaster same.

**Comment by mattgodbolt** [comment]
Yes :) It seems like it's fixed as part of the other timing fixes, which is
great!

We should try a much longer time and see if we still agree!

Hooray!

**Comment by mattgodbolt** [comment]
Actually; looking back on this it seems jsbeeb is still wrong compared to a master: I still get 305f at 0800.

**Comment by mattgodbolt** [comment]
We agree at 0x7e0 though (3032, 3032).
At 7e8 I get: 303c
At 7f0 I get: 3047
At 7f8 I get: 3050

I really wish my BBC Master wasn't in my loft in Horsham instead of here in Chicago!!

**Comment by BigEd** [comment]
At 7e8 I get: 303c  - confirmed
At 7f0 I get: 3047  - confirmed
At 7f8 I get: 3050  - confirmed
and then:
7f9 3054
7fa 3056
7fb 3056
7fc 3056
7fd 3057
7fe 3059
7ff 305D

**Comment by mattgodbolt** [comment]
Thanks Ed! More to check out on the train ride home!

**Comment by mattgodbolt** [comment]
So we agree on all those: it's the cycle at 800 that is the difference! I wonder what on earth can be causing that!

Real master is on ROL &7c; jsbeeb has made it to the following ROL A.

 Wonder if this is an IRQ-mid-instruction difference somehow...

**Comment by mattgodbolt** [comment]
Or an even/odd 1MHz bus initial setup difference somehow. Can't be that really as it's so reproducible.

**Comment by mattgodbolt** [comment]
I just disabled the check for IRQs happening on the penultimate cycle; making it on the last cycle... Still same results.

**Comment by mattgodbolt** [comment]
I tried fudging the IRQ time all over the shop, no joy. 

I put some breakpoints around. It's the second time we hit the 3057ish area that the timer goes to zero.

Putting the breakpoint a few cycles earlier yields: (NB the timer values here are multiplied up by two, ie in cpu cycles, )

    3057 c5 78.x CMP $78    ; t1c = 2
    3059 90 02.. BCC $305d ; t1c = -1 --- branch is taken
(    305b e5 78.x SBC $78)
    305d 26 7c&| ROL $7c  ; t1c = 0x4e1c (i.e the reloaded value), IRQ is pending...
    305f 2a*ROL A
    3060 c5 78.xCMP $78

so my hunch is it's a difference in when the IRQ is checked for *taken* branches. If so, this is an awesome find! My train's about to arrive home though, so sadly I can't investigate further just now...more on this later and/or tomorrow!

**Comment by mattgodbolt** [comment]
(cursory glance at branch code looks "sensible" but no real idea what the real 6502 does! some visual 6502 experimentation needed, assuming cmos and nmos don't differ in behaviour!)

**Comment by mattgodbolt** [comment]
no wait! branch code is weird...

was:
                cpu.pc = newPc;
                cpu.polltime(pageCrossed ? 3 : 1);
                cpu.checkInt();
                cpu.polltime(pageCrossed ? 1 : 2);

if I make it:
                cpu.pc = newPc;
                cpu.polltime(pageCrossed ? 3 : 2);
                cpu.checkInt();
                cpu.polltime(pageCrossed ? 1 : 1);

we agree! need to get off train argh! no idea if this breaks KE protection...

**Comment by mattgodbolt** [comment]
I win weirdest man in Evanston award. Am walking down road with laptop out running unit tests. All passing! I wonder what on earth that branch code was intending? Any ideas Rich?

Think we have it though! Need to engineer some page crossing branches too to be sure that's also right!

Nothing checked in yet. Yay though!

**Comment by mattgodbolt** [comment]
With the branch timings fix, here's some other sample points to compare against, should you have the time @BigEd (and again thanks!)

timer | A | PC
-------|----|------
1000 | f9 | 23eb
1800 | ed | c027
2000 | ef | c71e
3000 | ea | eaea
4000 | ea | eaea

clearly we need to run a bit longer to get useful results, so I then ran with 8 digits:

timer | A | PC
-------|----|------
1000 | f9 | 2f27
1800 | f9 | 242b
2000 | f9 | 21c4
2800 | e9 | e2d6

not quite sure what to make of it. Will try the original test again :)

**Comment by mattgodbolt** [comment]
1000 PI from the original disc no, FXes: 158.77 (vs hoglet's 160.51)
with FX16,0 : 157.96 (vs hoglet's 158.88)

We're getting there :)

**Comment by richtw1** [comment]
Yeah, we emulate that :)

**Comment by BigEd** [comment]
So this is some new tweak that's needed? That is an interesting find!

**Comment by richtw1** [comment]
I haven't really followed this in depth (I need to catch up properly 
with the thread) but does this timing discrepancy only affect the 
Master?  Is it possible that the branch logic which short-circuits T0 in 
the NMOS 6502s was changed in the CMOS versions?  I guess we don't 
really have any way of knowing at the moment, until a 65C02 is decapped 
and analysed.

**Comment by BigEd** [comment]
On the Master, 4 digits, we agree on the datapoints you tabulated.

Trying the 8 digit ones:
1000 - I get 2fc7 - maybe you made a typo because for me that matches the HEAD of jsbeeb
1800 - matched
2000 - matched
2800 - matched

Let's go for broke
timer of ff00
ask for 20 digits, result on Master is ED C70F
ask for 22 digits, result on Master is F9 300A
ask for 44 digits, result on Master is F9 2E60
ask for 99 digits, result on Master is F9 CC34
and jsbeeb HEAD agrees with all of those!

timer of 8000
20 digits, F9 2FD7
and again that's fine.

So here's a new twist, where jsbeeb and my master differ. Might need to prove that our BASIC ROMs are identical...
390RTS
443CALLG%
444FORI=0TO9STEP.1:X=3/ATN(1):N.:GOTO490
With timer preset to 8000, the master comes back with E0 A64E
and jsbeeb comes back with E0 A64B

I think the ROMs are the same, as they agree on
X%=0:FORI%=&8000TO&C000STEP4:X%=((X%AND&FFFFFF)*9)EOR!I%:N.:P.X%
returning 12737583

**Comment by BigEd** [comment]
Yes, it's a good guess that the C02 timings are somehow a little different. My Beeb is a bit poorly but I might be able to get some timings from it. My expectation is that it's cycle-accurate but it would be nice to confirm that's true even with the pi spigot benchmark.

**Comment by mattgodbolt** [comment]
Unless I made a mistake (will attach screenshots if I can get them to work ); we differ! I see f9/993f when I do what I think you're doing... But you tested with jsbeeb too! Wonder what's different here, I bet I've done something wrong on my side!

Can you elaborate on the suspected T0 differences? If you're talking branches those seem to match now...

**Comment by BigEd** [comment]
Here's a standalone disk and text version of what I think I ran.
[TBAS.ssd.gz](https://github.com/mattgodbolt/jsbeeb/files/378225/TBAS.ssd.gz)
[TBAS.txt](https://github.com/mattgodbolt/jsbeeb/files/378224/TBAS.txt)

**Comment by mattgodbolt** [comment]
Thanks @BigEd will try in a bit :)

**Comment by mattgodbolt** [comment]
Running TBAS gives me: `fb 8fbf`

I saved the disc (with TBAS as a BASIC program) in a shared google drive which we should all be able to edit and run...in theory: http://bbc.godbolt.org/beta/?model=Master&disc=gd%3A0B1Jaz13h1gYIX2VCOVVSOGg3RXc%2FTBAS_shared

**Comment by BigEd** [comment]
(Just seen your latest reply - might be invalid because of the following...)

Oops, I need to set C% before assembling, not after!
10C%=&8000
delete line 400
440P.~C%

Here are some results from TBAS
Timer 8000, Master result E0A653, JSBeeb HEAD result same.
Timer FF00, Master result EAA636, JSBeeb HEAD result same.
Timer 4000, Master result E0A655, JSBeeb HEAD result same.

So, I'm failing to find a discrepancy with this simpler case. Bother. Maybe I'm just doing everything wrong today!

I also reran the tweaked version of PIT as described above where I added in the ATN test, and now it too has the Master and JSBeeb in agreement. So, apologies, I think in conclusion I have not yet found a discrepancy with the latest code.

Although, if we don't yet agree on the centisecond level, it is possible that there's still something to be found.

Perhaps this bug has served it's purpose?

**Comment by mattgodbolt** [comment]
(this is where having #116 implemented would make life a lot easier!)

**Comment by mattgodbolt** [comment]
Thanks again @BigEd :) This whole thing is super confusing (I'm so glad you spotted my typo from last night!)... I think the bug is still valuable in terms of getting the centisecond stuff done.

If I make an SSD measuring tons of instructions, can you run it on a beeb and master for me? I think that's the only way we can be sure all the instructions are OK...then it's "just" IRQs and NMIs and other configuration-level stuff to worry about...

**Comment by BigEd** [comment]
Yes, I can have a go running an SSD, although I have to load binaries onto my Master manually - it has a serial connection but not an SSD capability. And my B is a bit flaky, but I'll do what I can!

**Comment by BigEd** [comment]
(I don't really understand what happened with the fix to the branch behaviour, because the code at the point of discrepancy was some way from a branch. Is that all explicable?)

**Comment by mattgodbolt** [comment]
Cool re: SSD; I might put together something :)

Re: branch behaviour, the discrepancy was indeed a few instructions from a branch...but the discrepancy happened on the destination of a taken branch! see the little ascii table I did above: the 3059 vs 305d case (I can elaborate if I'm not being clear here, as I suspect I'm not!)

**Comment by BigEd** [comment]
Ah yes, I can see it now - I was confused by my earlier disassembly which contained no branches, but of course was illustrating a quite different discrepancy.

**Comment by mattgodbolt** [comment]
No worries, this whole thing is super confusing :D

**Comment by richtw1** [comment]
StarDot thread which I believe confirms differences between NMOS/CMOS handling of interrupts in branches.
https://stardot.org.uk/forums/viewtopic.php?f=3&t=15631

Re https://github.com/mattgodbolt/jsbeeb/commit/f65a02d1677ba42ddddb50fd7847a6549a0bb5c8
The original behaviour was correct for NMOS, the changed behaviour is correct for CMOS.

**Comment by BigEd** [comment]
I think I can now test any given SSD on both Beeb and Master fairly readily, if need be.

**Comment by mattgodbolt** [comment]
Thanks both! Been totally rushed off my feet recently (new job and other
hobby project is busy as heck), if anyone else can take a look at this is
be delighted!

**Comment by scarybeasts** [comment]
Oh, breaking 6502 branch interrupt check timing in order to fix 65c12 branch interrupt check timing seems a bit vexatious. I'll send a PR that gets both correct.

@BigEd @mattgodbolt What else is pending in order to close this bug?

**Comment by scarybeasts** [comment]
PR to fix 6502 branch interrupt check timing is:
https://github.com/mattgodbolt/jsbeeb/pull/212

I've noticed some other bugs in the general area of BRK / IRQ / NMI etc., I'll crank through them as time permits.

**Comment by mattgodbolt** [comment]
Thanks @scarybeasts . In terms of remaining word, I'm not actually sure. We have the "pi spigot" test program and can try and use that to "prove" we run at the same speed as a real beeb, but I'm not sure how feasible that is. And if it doesn't match it's impossible to know what's differing. Though from @BigEd 's more recent (2018!) replies perhaps we're "done" on this bug now...

**Comment by BigEd** [comment]
It would be quite good to rerun any tests used in this thread, hoping for cycle accuracy and recording any discrepancies. Just possibly raising new issues for anything seen, to keep things as unconfusing as possible.  It might be that all is now well.

**Comment by scarybeasts** [comment]
A fairly random thought but does the Master sync elapsed TIME (as in the variable) with the Real Time Clock elapsed time? If so, perhaps old hardware drifts relative to real time? @BigEd

**Comment by scarybeasts** [comment]
On Tricky's Master: "Mode 7, 80 digits 1.08 3 out of 3, DFS all other ROMs UNPLUGed". This matches current jsbeeb at 1.08 so I think we're in good shape. We've improved various things under the umbrella of this bug, including interrupt vs. branch behavior.

Closing this bug for now. If additional issues remain or surface, we can file a fresh bug.

**Comment by mattgodbolt** [comment]
Amazing stuff: thanks @scarybeasts and Tricky!
